Solve multi-thearding bug

mathieuouillon opened the pull request:
Change Vector to CopyOnWriteArrayList safe-thread class

Commit:
Solve multi-thearding bug

Change Vector to CopyOnWriteArrayList safe-thread class

raffaelladevita commented:
The multi-threading issue was presumably caused by 
https://github.com/JeffersonLab/clas12-offline-software/blob/f0ae5e0671ee3b634aaf4919406a79fe5081b955/reconstruction/rtpc/src/main/java/org/jlab/rec/rtpc/KalmanFilter/KalmanFitter.java#L33-L34
where the material map is generated in the constructor of the KalmanFitter.

Since an instance of the KalmanFitter is created for every event, the material map is rebuilt too. Also, the material classes use Vectors that I understand can be accessed only by one thread at a time.

The modification in this pull request solves the multi-threading issue in accessing vectors by using the CopyOnWriteArrayList, which I believe creates a fresh copy of the list every time.

Since there is presumably no need to recreate the material map at every event, I think it would be better to move the generation of the material map to the init of the Engine, which is executed only once. Also, replacing Vector with ArrayList in the material classes would address the multi-threading issue and would not have drawbacks if the material map is created only once and made available to the KalmanFitter in the processEvent.

raffaelladevita commented:
For the future merging of this branch to the main branch, the following comments should be considered:
- it seems only a small part of the eloss and material-related classes and methods are used; if so, it would be better to keep only what is necessary;
- classes like Particle or PhysicalConstants already exist in common-tools and should not be duplicated but, if necessary, modified adding what is missing

Commit:
Remove old version

remove Material Class, Kalman Filter and Energy Loss calculation, to prepare to use the material class and the energy loss calculation from the common-tools

Commit:
Kalman Filter with the common-tools Material Class
